### PR TITLE
Implement light client and test code

### PR DESCRIPTION
**Pull request by hyeonLewis**
PR for implementing light client & Treasury contract.

### CODE REVIEW

File: contract/contracts/Library/Verify.sol [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+pragma solidity ^0.8.0;
+
+/**
+ * @dev Functions to verify signature
+ */
+library Verify {
+    /// @dev Since nested mapping makes struct and logic too complicated, we use a simple array to store validator set.
+    struct BlockHeader {
+        bytes author;
+        bytes[] prev_block_finalization_proof;
```
Comment (edwardycl) [inline contract/contracts/Library/Verify.sol]: Follow the solidity naming conventions.
Docs: https://docs.soliditylang.org/en/v0.8.17/style-guide.html#naming-conventions
```suggestion
        bytes[] prevBlockFinalizationProof;
```

File: contract/contracts/Library/Verify.sol [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+            block_header.timestamp > prev_block_header.timestamp,
+            "Verify::verify_header_to_header: Invalid block timestamp"
+        );
+
+        for (uint i = 0; i < prev_block_header.validators.length; i++) {
+            if (keccak256(prev_block_header.validators[i]) == keccak256(block_header.author)) {
+                break;
+            } else {
+                if (i == prev_block_header.validators.length - 1) {
+                    return false;
```
Comment (edwardycl) [inline contract/contracts/Library/Verify.sol]: Other conditions are handled by reverting, but checking the validator set is done by returning a boolean value. It could be made consistent by using a `revert` statement and removing the return value.

File: contract/contracts/Library/Verify.sol [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+        return true;
+    }
+
+    function verify_finalization_proof(
+        BlockHeader memory header,
+        bytes32 header_hash,
+        bytes[] memory finalization_proof
+    ) internal pure returns (bool) {
+        uint64 total_voting_power;
+        uint64 voted_voting_power;
```
Comment (edwardycl) [inline contract/contracts/Library/Verify.sol]: Use `uint256` to save gas and prevent overflows.

File: contract/contracts/Library/Verify.sol [hunk 4]
```diff
@@ -0,0 +1,163 @@
...
+        }
+        uint k = 0;
+        for (uint j = 0; j < finalization_proof.length; j++) {
+            (bytes memory signer, bytes memory signature) = abi.decode(
+                finalization_proof[j],
+                (bytes, bytes)
+            );
+            (bytes32 r, bytes32 s, uint8 v) = split_signature(signature);
+            if (pub_to_address(signer) == ecrecover(to_prefixed_hash(header_hash), v, r, s)) {
+                voted_voting_power += header.voting_power[k];
```
Comment (edwardycl) [inline contract/contracts/Library/Verify.sol]: Should check which validator is the signer of the proof. It is currently counting the voting power according to the validators array index.

File: contract/contracts/Treasury/EVMTreasury.sol [hunk 5]
```diff
@@ -124,22 +126,34 @@ contract EVMTreasury is Pausable, ReentrancyGuard, IEVMTreasury {
...
+        clients[block_header.block_height] = Client(
+            block_header.block_height,
+            header,
+            client.chain_name
         );
-        client.height += 1;
+        client.height = block_header.block_height;
         client.last_header = header;
-        clients[client.height] = client;
+        client.chain_name = client.chain_name;
```
Comment (edwardycl) [inline contract/contracts/Treasury/EVMTreasury.sol]: Redundant line of code.
```suggestion
```

File: contract/contracts/Treasury/EVMTreasury.sol [hunk 6]
```diff
@@ -124,22 +126,34 @@ contract EVMTreasury is Pausable, ReentrancyGuard, IEVMTreasury {
...
+    ) public whenNotPaused {
+        Verify.BlockHeader memory block_header = Verify.parse_header(header);
         require(
-            keccak256(abi.encodePacked(proof)) == keccak256(abi.encodePacked("valid")),
-            "EVMTreasury::update_light_client: Invalid block finalization proof"
+            Verify.verify_header_to_header(client.last_header, header),
+            "EVMTreasury::update_light_client: Invalid header"
+        );
+        require(
+            Verify.verify_finalization_proof(block_header, keccak256(header), proof),
```
Comment (edwardycl) [inline contract/contracts/Treasury/EVMTreasury.sol]: How about saving the proof in the `Client` storage and skip the re-verification process in the `verify_header_to_header` function?

**Review by edwardycl — CHANGES_REQUESTED** [review]

File: contract/contracts/Library/Verify.sol (hunk 3 above)
Comment (hyeonLewis) [reply contract/contracts/Library/Verify.sol:69]: I think rust struct also needs to be changed.

File: contract/contracts/Library/Verify.sol (hunk 4 above)
Comment (hyeonLewis) [reply contract/contracts/Library/Verify.sol]: Because it will always be the same for now, but for launching, as you said, we need to use ```validatorSet(bytes => uint64) or arrays ``` for checking validators to prevent attackers. (We need to talk about the implementation in discord)

File: contract/contracts/Library/Verify.sol (hunk 1 above)
Comment (hyeonLewis) [reply contract/contracts/Library/Verify.sol]: Fixed.

File: contract/contracts/Library/Verify.sol (hunk 2 above)
Comment (hyeonLewis) [reply contract/contracts/Library/Verify.sol]: Fixed.

File: contract/contracts/Library/Verify.sol (hunk 3 above)
Comment (hyeonLewis) [reply contract/contracts/Library/Verify.sol]: Fixed.

File: contract/contracts/Treasury/EVMTreasury.sol (hunk 5 above)
Comment (hyeonLewis) [reply contract/contracts/Treasury/EVMTreasury.sol]: Fixed.

File: contract/contracts/Treasury/EVMTreasury.sol (hunk 6 above)
Comment (hyeonLewis) [reply contract/contracts/Treasury/EVMTreasury.sol]: What does the re-verification process mean?

File: contract/contracts/Library/Verify.sol (hunk 2 above)
Comment (edwardycl) [reply contract/contracts/Library/Verify.sol]: The boolean return value should be removed. It makes users to think that this function uses a return value for error handling

File: contract/contracts/Treasury/EVMTreasury.sol (hunk 6 above)
Comment (edwardycl) [reply contract/contracts/Treasury/EVMTreasury.sol]: The current block proof is re-verified when the next block is updated, in the `verifyHeaderToHeader` function. Is there a chance for the proof to be updated at the next block?

Comment (hyeonLewis) [reply contract/contracts/Treasury/EVMTreasury.sol]: Yes, but it seems intended. You can check [here](https://github.com/postech-dao/simperby/blob/main/common/src/light_client.rs)

File: contract/contracts/Library/Verify.sol (hunk 2 above)
Comment (hyeonLewis) [reply contract/contracts/Library/Verify.sol]: My bad, thanks.

File: contract/contracts/Library/Verify.sol (hunk 4 above)
Comment (hyeonLewis) [reply contract/contracts/Library/Verify.sol]: Fixed.

File: contract/contracts/Library/Strings.sol [hunk 7]
```diff
@@ -0,0 +1,83 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.0;
+
+/**
+ * @dev Functions about strings.
+ */
+library Strings {
+    function bytesToString(bytes memory byteCode) internal pure returns (string memory stringData) {
```
Comment (edwardycl) [inline contract/contracts/Library/Strings.sol]: `bytes` to `string` conversion could be done with `string(byteCode)`

File: contract/contracts/Library/Verify.sol [hunk 8]
```diff
@@ -0,0 +1,361 @@
...
+    uint constant uint128Length = 16;
+    uint constant strUint64Length = 8;
+    uint constant enumLength = 4;
+
+    struct TypedSignature {
+        bytes signature;
+        bytes signer;
+    }
+
+    struct validatorSet {
```
Comment (edwardycl) [inline contract/contracts/Library/Verify.sol]:
```suggestion
    struct ValidatorSet {
```

File: contract/contracts/Treasury/EVMTreasury.sol [hunk 9]
```diff
@@ -3,102 +3,147 @@ pragma solidity ^0.8.0;
...
-        address indexed receiver_address,
-        uint256 contract_sequence
+        address indexed tokenAddress,
+        uint256 tokenIndex,
+        address indexed receiverAddress,
+        uint256 contractSequence
     );
 
-    event UpdateLightclient(uint256 indexed height, string last_header, string chain_name);
+    event UpdateLightClient(bytes indexed lastHeader);
```
Comment (edwardycl) [inline contract/contracts/Treasury/EVMTreasury.sol]: Why did you remove the block height?

File: contract/contracts/Treasury/EVMTreasury.sol [hunk 10]
```diff
@@ -3,102 +3,147 @@ pragma solidity ^0.8.0;
...
 
-    Client public client;
-
-    mapping(uint256 => Client) public clients;
+    LightClient public lightClient;
 
+    mapping(uint256 => LightClient) public lightClients;
     /* ========== EVENTS ========== */
 
     event TransferFungibleToken(
```
Comment (edwardycl) [inline contract/contracts/Treasury/EVMTreasury.sol]: Events are usually defined in the interface. Move the definitions to `IEVMTreasury.sol`

File: contract/contracts/Treasury/EVMTreasury.sol [hunk 11]
```diff
@@ -110,42 +155,52 @@ contract EVMTreasury is Pausable, ReentrancyGuard, IEVMTreasury {
...
+
+        lightClient.lastHeader = header;
+        lightClient.repositoryRoots.push(_blockHeader.repositoryMerkleRoot);
+        lightClient.commitRoots.push(_blockHeader.commitMerkleRoot);
+
+        lightClients[_blockHeader.blockHeight] = LightClient(
+            lightClient.heightOffset,
+            header,
+            lightClient.repositoryRoots,
+            lightClient.commitRoots
```
Comment (edwardycl) [inline contract/contracts/Treasury/EVMTreasury.sol]: This would result in a O(N^2) scale of storage usage and the stored merkle roots are redundant.
How about only saving the block headers?

File: contract/contracts/Library/Verify.sol [hunk 12]
```diff
@@ -0,0 +1,361 @@
...
+        bytes version;
+    }
+
+    /* ========== VERIFY FUNCTIONS ========== */
+    /**
+     * @dev Verifies new header is valid.
+     * @param prevHeader Bytes of previous header (Current lastHeader).
+     * @param header New header from relayer (Candidate for new lastHeader).
+     */
+    function verifyHeaderToHeader(bytes memory prevHeader, bytes memory header) internal pure {
```
Comment (edwardycl) [inline contract/contracts/Library/Verify.sol]: Take `BlockHeader` structs as parameters to separate the parsing and the verification.

File: contract/contracts/Library/Verify.sol [hunk 13]
```diff
@@ -0,0 +1,361 @@
...
+    ) internal pure {
+        uint256 _totalVotingPower;
+        uint256 _votedVotingPower;
+        for (uint i = 0; i < header.validators.length; i++) {
+            _totalVotingPower += header.validators[i].votingPower;
+        }
+        uint k = 0;
+        for (uint j = 0; j < finalizationProof.length; j++) {
+            (bytes32 r, bytes32 s, uint8 v) = splitSignature(finalizationProof[j].signature);
+            if (Utils.pkToAddress(finalizationProof[j].signer) == ecrecover(headerHash, v, r, s)) {
```
Comment (edwardycl) [inline contract/contracts/Library/Verify.sol:113]: Identify the signer, just in case.
```suggestion
            if (Utils.pkToAddress(finalizationProof[j].signer) == ecrecover(headerHash, v, r, s)) {
                require(finalizationProof[j].signer == header.validators[k].validator);
```

**Review by edwardycl — CHANGES_REQUESTED** [review]

File: contract/contracts/Treasury/EVMTreasury.sol [hunk 14]
```diff
@@ -3,102 +3,147 @@ pragma solidity ^0.8.0;
...
     string public constant name = "EVM SETTLEMENT CHAIN TREASURY V1";
+    bytes public constant chainName = hex"6d797468657265756d"; // mythereum, for testing
+    uint128 public constant contractSequence = 0;
 
-    Client public client;
-
-    mapping(uint256 => Client) public clients;
+    LightClient public lightClient;
 
+    mapping(uint256 => LightClient) public lightClients;
```
Comment (edwardycl) [inline contract/contracts/Treasury/EVMTreasury.sol]: Add public variables to the interface.
i.e. `function lightClient() returns (LightClient)`

Comment (hyeonLewis) [reply contract/contracts/Treasury/EVMTreasury.sol]: Actually, we need a separate view function for commitRoots. I will add both.

File: contract/contracts/Library/Strings.sol (hunk 7 above)
Comment (hyeonLewis) [reply contract/contracts/Library/Strings.sol]: You were right, fixed.

File: contract/contracts/Library/Verify.sol (hunk 8 above)
Comment (hyeonLewis) [reply contract/contracts/Library/Verify.sol]: Fixed.

File: contract/contracts/Treasury/EVMTreasury.sol (hunk 9 above)
Comment (hyeonLewis) [reply contract/contracts/Treasury/EVMTreasury.sol]: Temporarily removed for testing, fixed.

File: contract/contracts/Treasury/EVMTreasury.sol (hunk 10 above)
Comment (hyeonLewis) [reply contract/contracts/Treasury/EVMTreasury.sol]: Fixed.

File: contract/contracts/Treasury/EVMTreasury.sol (hunk 11 above)
Comment (hyeonLewis) [reply contract/contracts/Treasury/EVMTreasury.sol]: As we talked, I deleted it. Fixed.

File: contract/contracts/Library/Verify.sol (hunk 12 above)
Comment (hyeonLewis) [reply contract/contracts/Library/Verify.sol]: Fixed.

File: contract/contracts/Library/Verify.sol (hunk 13 above)
Comment (hyeonLewis) [reply contract/contracts/Library/Verify.sol:113]: Added.

File: contract/contracts/Treasury/EVMTreasury.sol (hunk 14 above)
Comment (hyeonLewis) [reply contract/contracts/Treasury/EVMTreasury.sol]: Added.

Comment (edwardycl) [reply contract/contracts/Treasury/EVMTreasury.sol]: I think `contractSequence` could also be the query target. (and maybe `name` and `chainName`)

Comment (hyeonLewis) [reply contract/contracts/Treasury/EVMTreasury.sol]: All targets(public variables) are added.

**Review by edwardycl — APPROVED** [review]
👍
